### PR TITLE
feat: Update to the new way of threading messages in Google Chat

### LINKED ISSUE #40: sameThreadNotification doesn't work
### Jenkins and plugins versions report

The [`sameThreadNotification`](https://github.com/jenkinsci/google-chat-notification-plugin#:~:text=for%20card%20message.-,sameThreadNotification,-This%20is%20an) parameter for the `googlechatnotification` command doesn't work: All messages appear in the chat as new threads.

The plugin is using the `threadKey` query parameter in the URL. When that parameter is used, a new thread will be created unless the [`messageReplyOption`](https://developers.google.com/chat/api/reference/rest/v1/spaces.messages/create#query-parameters) parameter is also used. Simplest is to always use `&messageReplyOption=REPLY_MESSAGE_FALLBACK_TO_NEW_THREAD` whenever `threadKey` is used. That will automatically create a new thread if none exists for that key, or create a reply to an existing thread.

However, although the `threadKey` parameter works when `messageReplyOption` is used, using `threadKey` as a query parameter is now deprecated. The preferred method is to include [thread information](https://developers.google.com/chat/api/reference/rest/v1/spaces.messages#Thread) in the message JSON.

### What Operating System are you using (both controller, and any agents involved in the problem)?

Linux controller and agent, but OS isn't relevant to this problem.

### Reproduction steps

Send two messages with sameThreadNotification enabled:

```groovy
googlechatnotification url: 'id:example-webhook', messageFormat: 'simple', sameThreadNotification: 'true', message: 'message from pipeline'
googlechatnotification url: 'id:example-webhook', messageFormat: 'simple', sameThreadNotification: 'true', message: 'This should be a reply to the previous message.'
```

### Expected Results

Second message is a threaded reply.
image

### Actual Results

Both messages are new threads.
image

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Ryudo302**
Following the instructions from #40 (thanks to @gpflaum).

Resolves #40

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Review by gpflaum — COMMENTED** [review]
The code changes look good, except that ` sameThreadNotification: 'true'` will have no effect with `messageFormat: 'card'`. The user could set their own threadKey in the message json they pass to `googlechatnotification`. If the `sameThreadNotification` and `threadKey` parameters won't work for card messages, the documentation should be updated.

I'll try to test this PR on Monday or Tuesday.

**Review by gpflaum — COMMENTED** [review]
Threading didn't work for me. It looks like the [messageReplyOption](https://developers.google.com/chat/api/reference/rest/v1/spaces.messages/create#query-parameters) query parameter is required for this. Only the `threadKey` query parameter is deprecated. 

Here's a working example:
```sh
export GCHAT_WEBHOOK_URL="https://chat.googleapis.com/v1/spaces/space-redacted/messages?key=key-redacted&token=token-redacted"

curl -X POST \
  -H 'Content-Type: application/json; charset=UTF-8' \
  -d '{"text":"message from curl", "thread":{"threadKey":"threading-test-02"}}' \
  "${GCHAT_WEBHOOK_URL}&messageReplyOption=REPLY_MESSAGE_FALLBACK_TO_NEW_THREAD"

curl -X POST \
  -H 'Content-Type: application/json; charset=UTF-8' \
  -d '{"text":"reply from curl", "thread":{"threadKey":"threading-test-02"}}' \
  "${GCHAT_WEBHOOK_URL}&messageReplyOption=REPLY_MESSAGE_FALLBACK_TO_NEW_THREAD"
```

**Comment by Ryudo302** [comment]
Hi, @gpflaum!

I apologize for the delay in resuming the work with this issue.

I found that the problem here is specifically to the in-line threaded spaces, which became the default behavior since March this year: [All newly created spaces in Google Chat will be in-line threaded beginning March 13, 2023](http://workspaceupdates.googleblog.com/2023/02/new-google-chat-spaces-will-be-in-line-threaded.html). This issue don't affect spaces organized by topics.

Anyway, I think that now the problem is solved. I've made some basic tests with a Google Chat Mock Server. May you please test again?

File: src/main/java/io/cnaik/service/ResponseMessageUtil.java [hunk 1]
```diff
@@ -31,7 +36,34 @@ public ResponseMessageUtil(GoogleChatNotification googleChatNotification) {
     }
 
     public String createTextMessage() {
-        return escapeSpecialCharacter(replaceJenkinsKeywords(googleChatNotification.getMessage()));
+        JSONObject thread = null;
+        if (googleChatNotification.isSameThreadNotification()) {
+            thread = new JSONObject();
+
+            var threadKey = StringUtils.defaultIfBlank(googleChatNotification.getThreadKey(), getJobName());
```
[the remark this answers is not in the document]
Comment (Ryudo302) [reply src/main/java/io/cnaik/service/ResponseMessageUtil.java]: @gpflaum, could you open a new issue about this, please? Maybe requesting this configuration to be parameterized.

**Comment by Ryudo302** [comment]
I've made some tests at my job's Google Chat spaces and apparently everything is now working!

**Pull request merged**